Source/bmalloc/libpas/src/libpas/pas_report_crash.c

```diff
@@ -97,57 +110,68 @@ kern_return_t pas_report_crash_extract_pgm_failure(vm_address_t fault_address, m
-            report->fault_address = fault_address;
-            report->allocation_size = pgm_metadata->allocation_size_requested;
-            return KERN_SUCCESS;
-        }
+        if (pgm_metadata->right_align) {
+            report->alignment = "address right-aligned";
+
+            // Right-aligned "Lower PGM OOB" checking
+            addr64_t bottom = (addr64_t) (key - pgm_metadata->mem_to_waste - pgm_metadata->page_size);
+            addr64_t top = (addr64_t) (key - pgm_metadata->mem_to_waste);
+
+            if (pas_fault_address_is_in_bounds(fault_address, bottom, top))
+                pas_update_report_crash_fields(report, "out-of-bounds", "high", fault_address, pgm_metadata->allocation_size_requested);
+
+            // Right-aligned "UAF + OOB" Check towards lower guard page
+            bottom = (addr64_t) (key - pgm_metadata->mem_to_waste - pgm_metadata->page_size);
+            top = (addr64_t) key;
+
+            if (pas_fault_address_is_in_bounds(fault_address, bottom, top))
+                pas_update_report_crash_fields(report, "OOB + UAF", "low", fault_address, pgm_metadata->allocation_size_requested);
+
+            // "Upper PGM OOB" Checking
+            // Address range for "UAF + OOB" towards upper guard page == "upper pgm OOB" address range
+            // Which type of access is categorized depends on memory aliveness
+            // Currently as metadata is accessible its considered "alive" so categorized "OOB" access type
+            bottom = (addr64_t) (key - pgm_metadata->mem_to_waste + pgm_metadata->size_of_data_pages);
+            top = (addr64_t) (key - pgm_metadata->mem_to_waste + pgm_metadata->size_of_data_pages + pgm_metadata->page_size);
+
+            if (pas_fault_address_is_in_bounds(fault_address, bottom, top))
+                pas_update_report_crash_fields(report, "out-of-bounds", "high", fault_address, pgm_metadata->allocation_size_requested);
+
+        } else {
+            report->alignment = "address left-aligned";
+
+            // Left-aligned "Lower PGM OOB" checking
+            // Address range for "UAF + OOB" towards lower guard page == "lower pgm OOB" address range
+            // Which type of access is categorized depends on memory aliveness
+            // Currently as metadata is accessible its considered "alive" so categorized "OOB" access type
+            addr64_t bottom = key - pgm_metadata->page_size;
+            addr64_t top = key;
+
+            if (pas_fault_address_is_in_bounds(fault_address, bottom, top))
+                pas_update_report_crash_fields(report, "out-of-bounds", "high", fault_address, pgm_metadata->allocation_size_requested);
+
+            // Left-aligned "UAF + OOB" check towards upper guard page
+            bottom = (addr64_t) (key + pgm_metadata->size_of_data_pages - pgm_metadata->mem_to_waste);
+            top = (addr64_t) (key + pgm_metadata->size_of_data_pages - pgm_metadata->mem_to_waste + pgm_metadata->page_size);
```

Comment:
This should check the "wasted memory" region:

```suggestion
            bottom = (addr64_t) (key + pgm_metadata->size_of_data_pages - pgm_metadata->mem_to_waste);
            top = (addr64_t) (key + pgm_metadata->size_of_data_pages);
```